### DISCUSSION TITLE
ImageDataset: accept list of lists of str as image_files

### DISCUSSION THREAD

**Discussion by grudloff** [Project-MONAI/tutorials#1517] [General]
I am fairly new to monai, so keep that in mind.

I am working with a classification task. The inputs are volumes (Height, Width, Depth), that are loaded by passing folder paths.

The classification are done over a patient, which can have multiple sessions (multiple volumes). Given the constraint of ImageDataset only accepting a single path for each label, the dataset obtained with the code bellow points to training the model on a per session manner.

```python
# paths: List[str] of length N
# labels: nd.array of shape (N, C)
# N: Number of unique patient_id, session_id pairs
# C: Number of classes of label array (one-hot-encoded)

check_ds = ImageDataset(image_files=paths, labels=labels)

im, label = check_ds[0]
# im: metatensor of shape (1, H,W,D)
# label: ndarray of shape (1, C)
```

Is there an straightforward way in monai to construct a dataset that accepts as input a list of lists of strs, and outputs a list of tensors? Or should I just build my own logic with a custom dataset that internally uses LoadImage?

In summary, I need a CustomDataset like this:

```python
# paths: List[List[str]] of length M
# labels: nd.array of shape (M, C)
# M: Number of unique patient_id
# C: Number of classes of label array (one-hot-encoded)

check_ds = CustomDataset(image_files=paths, labels=labels)

im, label = check_ds[0]
# im: List[metatensor]
# label: ndarray of shape (1, C)
```

**Comment by grudloff** [comment]
I am thinking about something like this:

```python
import numpy as np
import monai
from monai.data import Dataset
from monai.transforms import LoadImage, Compose, Spacing, EnsureChannelFirst

class CustomDataset(Dataset):
    def __init__(self, image_files, labels, transform=None):
        self.image_files = image_files
        self.labels = labels
        self.transform = transform

    def __len__(self):
        return len(self.image_files)

    def __getitem__(self, index):
        img_paths = self.image_files[index]
        img_list = [self.transform(path) for path in img_paths]
        label = self.labels[index]
        return img_list, label


transform = Compose([LoadImage(reader="PydicomReader", image_only=True, dtype=np.float32),
                     EnsureChannelFirst(channel_dim='no_channel'),
                     Spacing(pixdim=[RESIZE_DIM]*3),
                    ])

check_ds = CustomDataset(image_files=paths, labels=labels, transform=transform)

im, label = check_ds[0]
print(type(im), [i.shape for i in im], label, label.shape)
```

Any advise or more elegant solution would be appreciated!

**Comment by KumoLiu** [comment]
Hi @grudloff, I think you can directly use `Dataset` which can accept a list of data paths or a list of dictionaries.
https://github.com/Project-MONAI/MONAI/blob/5fd23b8c6a5bcb378063b99523af89c81880060c/monai/data/dataset.py#L64

Hope it helps, thanks!

**Reply by grudloff** [reply]
Yes! but I got and error. I noticed the issue was that data is concatenated at the end. In my case data is not necesarily uniform in shape or in spatial dimension, so it would require resizing to a common shape. That's why I was leaning towards a list of tensors as output, instead of stacked tensors along a channel dim.

**Reply by KumoLiu** [reply]
I think data should always be concatenated along the batch dim?

**Reply by grudloff** [reply]
I just found out that `PadListDataCollate` does just what I need to allow the concatenation, I will give that a try!

In any case, I was looking a the documentation of `ImageDataset` and I think it doesn't reflect clearly this behaviour. I would be up for doing a small PR and improve that. Let me know what you think

**Reply by grudloff** [reply]
Would you mind helping me out? I got two issues.

Here is the code:

```python
import torch
from monai.data import ImageDataset, DataLoader
from monai.transforms import Compose, Spacing, EnsureChannelFirst, PadListDataCollate

transform = Compose([
    EnsureChannelFirst(),
    Spacing(pixdim=[RESIZE_DIM]*3),
    PadListDataCollate()
])

check_ds = ImageDataset(image_files=paths, labels=labels, reader="PydicomReader", transform=transform,
                        dtype=np.float32)

im, label = check_ds[0]
print(type(im), im.shape, label, label.shape)
```

1. If the loaded instance has only one volume, it throws `NonImplementedError` because `collate_meta_tensor` expects a `Sequence`.
As a solution here I am thinking of building a transform that checks if the first dimension is 1, then just return the tensor as is.
2. Else, it throws an error because the affine matrices are not the same.
Here I am at a loss, I am not sure what the affine matrix is in this context

**Reply by grudloff** [reply]
Here is the solution for the 1st problem:
```python
from collections.abc import Sequence

class CustomPadListDataCollate(PadListDataCollate):
    def __call__(self, batch):
        if not isinstance(batch, Sequence):
            return batch
        return super().__call__(batch)
```

**Reply by KumoLiu** [reply]
Hi @grudloff, `PadListDataCollate` should accept a batch of data, so it always use in `DataLoader` instead of `Dataset`. If you only want to pad your data, you can directly use `SpatialPad`.
https://github.com/Project-MONAI/MONAI/blob/5fd23b8c6a5bcb378063b99523af89c81880060c/monai/transforms/croppad/batch.py#L72-L76

**Reply by grudloff** [reply]
You mean something like this?

```python
import torch
from monai.data import ImageDataset, DataLoader
from monai.transforms import Compose, Spacing, EnsureChannelFirst, PadListDataCollate

transform = Compose([
    EnsureChannelFirst(),
    Spacing(pixdim=[RESIZE_DIM]*3),
])

# paths: List[List[str]] of length M
# labels: nd.array of shape (M, C)
# M: Number of unique patient_id
# C: Number of classes of label array (one-hot-encoded)

check_ds = ImageDataset(image_files=paths, labels=labels, reader="PydicomReader", transform=transform,
                        dtype=np.float32)
check_loader = DataLoader(check_ds, collate_fn = PadListDataCollate())

im, label = monai.utils.misc.first(check_loader)
print(type(im), im.shape, label)
```

This looks right, but again throws an affine matrix error. The full trace stack is below

<details>
<summary>Error message</summary>
```bash
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[36], line 4
      1 #os.mkdir(OUT_PATH)
      2 
      3 # Iterate through the dataset and save preprocessed data to disk
----> 4 for idx, (im, label) in enumerate(check_loader):
      5     img_list, label = check_ds[idx]
      6     data = {'images': img_list, 'label': label}

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:634, in _BaseDataLoaderIter.__next__(self)
    631 if self._sampler_iter is None:
    632     # TODO([https://github.com/pytorch/pytorch/issues/76750)](https://github.com/pytorch/pytorch/issues/76750)%3C/span%3E)
    633     self._reset()  # type: ignore[call-arg]
--> 634 data = self._next_data()
    635 self._num_yielded += 1
    636 if self._dataset_kind == _DatasetKind.Iterable and \
    637         self._IterableDataset_len_called is not None and \
    638         self._num_yielded > self._IterableDataset_len_called:

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:678, in _SingleProcessDataLoaderIter._next_data(self)
    676 def _next_data(self):
    677     index = self._next_index()  # may raise StopIteration
--> 678     data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    679     if self._pin_memory:
    680         data = _utils.pin_memory.pin_memory(data, self._pin_memory_device)

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py:51, in _MapDatasetFetcher.fetch(self, possibly_batched_index)
     49         data = self.dataset.__getitems__(possibly_batched_index)
     50     else:
---> 51         data = [self.dataset[idx] for idx in possibly_batched_index]
     52 else:
     53     data = self.dataset[possibly_batched_index]

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py:51, in <listcomp>(.0)
     49         data = self.dataset.__getitems__(possibly_batched_index)
     50     else:
---> 51         data = [self.dataset[idx] for idx in possibly_batched_index]
     52 else:
     53     data = self.dataset[possibly_batched_index]

File /opt/conda/lib/python3.10/site-packages/monai/data/image_dataset.py:108, in ImageDataset.__getitem__(self, index)
    106 # load data and optionally meta
    107 if self.image_only:
--> 108     img = self.loader(self.image_files[index])
    109     if self.seg_files is not None:
    110         seg = self.loader(self.seg_files[index])

File /opt/conda/lib/python3.10/site-packages/monai/transforms/io/array.py:290, in LoadImage.__call__(self, filename, reader)
    282     raise RuntimeError(
    283         f"{self.__class__.__name__} cannot find a suitable reader for file: {filename}.\n"
    284         "    Please install the reader libraries, see also the installation instructions:\n"
    285         "    [https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies.](https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies.%3C/span%3E%3Cspan) class="ansi-bold" style="color:rgb(175,95,0)">\n"
    286         f"   The current registered: {self.readers}.\n{msg}"
    287     )
    289 img_array: NdarrayOrTensor
--> 290 img_array, meta_data = reader.get_data(img)
    291 img_array = convert_to_dst_type(img_array, dst=img_array, dtype=self.dtype)[0]
    292 if not isinstance(meta_data, dict):

File /opt/conda/lib/python3.10/site-packages/monai/data/image_reader.py:617, in PydicomReader.get_data(self, data)
    612         metadata[MetaKeys.ORIGINAL_CHANNEL_DIM] = self.channel_dim
    613     metadata["spacing"] = affine_to_spacing(
    614         metadata[MetaKeys.ORIGINAL_AFFINE], r=len(metadata[MetaKeys.SPATIAL_SHAPE])
    615     )
--> 617     _copy_compatible_dict(metadata, compatible_meta)
    619 return _stack_images(img_array, compatible_meta), compatible_meta

File /opt/conda/lib/python3.10/site-packages/monai/data/image_reader.py:128, in _copy_compatible_dict(from_dict, to_dict)
    126 affine_key, shape_key = MetaKeys.AFFINE, MetaKeys.SPATIAL_SHAPE
    127 if affine_key in from_dict and not np.allclose(from_dict[affine_key], to_dict[affine_key]):
--> 128     raise RuntimeError(
    129         "affine matrix of all images should be the same for channel-wise concatenation. "
    130         f"Got {from_dict[affine_key]} and {to_dict[affine_key]}."
    131     )
    132 if shape_key in from_dict and not np.allclose(from_dict[shape_key], to_dict[shape_key]):
    133     raise RuntimeError(
    134         "spatial_shape of all images should be the same for channel-wise concatenation. "
    135         f"Got {from_dict[shape_key]} and {to_dict[shape_key]}."
    136     )

RuntimeError: affine matrix of all images should be the same for channel-wise concatenation. Got [[  -0.896484    0.          0.        245.928242]
 [   0.         -0.896484    0.        220.488242]
 [   0.          0.          3.       -705.409   ]
 [   0.          0.          0.          1.      ]] and [[  -0.976563     0.           0.         253.0732815]
 [   0.          -0.976563     0.         237.0882815]
 [   0.           0.          -3.        -172.206    ]
 [   0.           0.           0.           1.       ]].
```
<\details>

**Reply by KumoLiu** [reply]
Yes, could you share a case and then I can help take a deep look?

**Reply by grudloff** [reply]
Sure! Thanks a lot for the help!

Here is a [kaggle notebook](https://www.kaggle.com/gabrielrudloff/monai-dicom) with the minimum required code.

**Reply by grudloff** [reply]
Hello @KumoLiu, I just added a new version to the notebook with a section at the end with an implementation that achieves what I wanted.

**Reply by grudloff** [reply]
I am still puzzled about `pad_list_data_collate` not working. In my transform, I normalize the spacing of each volume, shouldn't this make the affine transform the same? Sorry if I am missing something, as I said, I am new to monai and also to working with this kind of data.

**Reply by KumoLiu** [reply]
Hi @grudloff, I think the reason is that you have nested list in your paths.
![Screen Shot 2023-09-18 at 11 01 42](https://github.com/Project-MONAI/tutorials/assets/55491388/c6d47bcc-de15-4bce-b40e-7361057d1e20)

And the spacing you have in the nested list are different so when they stack, there will be error throw out.
https://github.com/Project-MONAI/MONAI/blob/281cb0119c01eaa8e6c841880b91f92f45e8d7f7/monai/transforms/io/array.py#L247

Hope it helps, thanks!
